lib/gtfs_realtime_viz.ex

```diff
@@ -56,13 +61,70 @@ defmodule GTFSRealtimeViz do
   @spec visualize_diff(term, term, route_opts) :: String.t
   def visualize_diff(group_1, group_2, opts) do
     routes = Map.keys(opts)
-    archive_1 = get_vehicle_archive(group_1, routes)
-    archive_2 = get_vehicle_archive(group_2, routes)
-    [vehicle_archive: Enum.zip(archive_1, archive_2), routes: opts, render_diff?: true]
+    vehicle_archive_1 = get_vehicle_archive(group_1, routes)
+    trip_archive_1 = get_trip_update_archive(group_1, routes)
+    vehicle_archive_2 = get_vehicle_archive(group_2, routes)
+    trip_archive_2 = get_trip_update_archive(group_2, routes)
+
+    [trip_update_archive: Enum.zip(trip_archive_1, trip_archive_2), vehicle_archive: Enum.zip(vehicle_archive_1, vehicle_archive_2), routes: opts, render_diff?: true]
     |> gen_html()
     |> Phoenix.HTML.safe_to_string()
   end
 
+  defp get_trip_update_archive(group, routes) do
+    group
+    |> State.trip_updates
+    |> trips_we_care_about(routes)
+    |> trip_updates_by_stop_id
+  end
+
+  def trips_we_care_about(state, routes) do
+    Enum.map(state,
+      fn {descriptor, update_list} ->
+        filtered_positions = update_list
+        |> Enum.filter(fn trip_update ->
+          trip_update.trip.route_id in routes
+        end)
+        {descriptor, filtered_positions}
+      end)
+  end
+
+  defp trip_updates_by_stop_id(state) do
+    Enum.map(state, fn {_descriptor, trip_updates} ->
+      trip_updates
+      |> Enum.flat_map(fn trip_update ->
+        trip_update.stop_time_update
+        |> Enum.reduce(%{}, fn stop_update, stop_update_acc ->
+          if stop_update.arrival do
+            Map.put(stop_update_acc, stop_update.stop_id, stop_update.arrival.time)
+          else
+            stop_update_acc
+          end
+        end)
+      end)
+    end)
+    |> Enum.map(fn predictions ->
+      Enum.reduce(predictions, %{}, fn {stop_id, time}, acc ->
+        Map.put(acc, stop_id, ([acc[stop_id], timestamp(time)] |> List.flatten |> Enum.reject(& &1 == nil)))
```

Comment:
Hm, I think this is building up a map where each key is a stop id, and each value is a list of times? Is this the same thing?

``` ex
Enum.reduce(predictions, %{}, fn {stop_id, time}, acc ->
  Map.update(acc, stop_id, [timestamp(time)], fn timestamps -> timestamps ++ [timestamp(time)] end)
end)
```

I think that's a bit more idiomatic - `Map.update` with a default value, rather than lists within lists, flattening, and rejecting `nil`.

Reply:
yeah, that seems better.